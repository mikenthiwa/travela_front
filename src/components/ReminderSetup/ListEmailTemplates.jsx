import React, {Component, Fragment} from 'react';
import PropTypes from 'prop-types';
import WithLoadingEmailTemplatesTable from './EmailTemplatesTable';
import TemplatesPagination from './TemplatesPagination';
import NoTemplates from './NoTemplates';
import TemplateDetailsModal from './TemplateDetailsModal';

class ListEmailTemplates extends Component {
  constructor(props) {
    super(props);
  }

  componentDidMount() {
    const {fetchTemplates, location: {search}} = this.props;
    fetchTemplates(search);
  }

  onPageChange = (previousOrNext) => {
    const {
      fetchTemplates, listEmailTemplatesReducer: {pagination: {currentPage}},
      location: {search}
    } = this.props;
    fetchTemplates(`${search}${search ? '&' : '?'}page=${previousOrNext === 'previous' ? currentPage - 1 : currentPage + 1}`);
  };

  renderPagination = (pageCount, currentPage) => {
    return (
      <TemplatesPagination
        onPageChange={this.onPageChange}
        pageCount={pageCount ? pageCount : 1}
        currentPage={currentPage ? currentPage : 1}
      />
    );
  };

  render() {
    const {
      listEmailTemplatesReducer: {templates, pagination, selectedTemplate},
      openModal, closeModal, modalType, shouldOpen,
      fetchOneTemplate, setItemToDisable, history, isLoading
    } = this.props;
    const {pageCount, currentPage} = pagination;
    return (
      <Fragment>
        <WithLoadingEmailTemplatesTable
          templates={templates}
          setItemToDisable={setItemToDisable}
          openModal={openModal}
          fetchOneTemplate={fetchOneTemplate}
          history={history}
          isLoading={isLoading}
        />
        {templates.length > 0 && !isLoading ? (
          <div>
            <TemplateDetailsModal
              closeModal={closeModal} modalType={modalType}
              shouldOpen={shouldOpen} templates={templates}
              history={history}
              selectedTemplate={selectedTemplate} />
            {this.renderPagination(pageCount,currentPage)}
          </div>
        ) :
          ('')}
      </Fragment>
    );
  }
}

ListEmailTemplates.propTypes = {
  fetchTemplates: PropTypes.func.isRequired, 
  listEmailTemplatesReducer: PropTypes.object.isRequired,
  location: PropTypes.object.isRequired,
  openModal: PropTypes.func.isRequired,
  closeModal: PropTypes.func,
  shouldOpen: PropTypes.bool,
  modalType: PropTypes.string,
  fetchOneTemplate: PropTypes.func,
  setItemToDisable: PropTypes.func.isRequired,
  history: PropTypes.object.isRequired,
  isLoading: PropTypes.bool
};

ListEmailTemplates.defaultProps = {
  modalType: '',
  shouldOpen: false,
  isLoading: false,
  fetchOneTemplate: () => {},
  closeModal: () => { }
};

export default ListEmailTemplates;
