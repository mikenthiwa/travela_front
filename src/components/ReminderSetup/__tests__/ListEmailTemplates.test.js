import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { mount } from 'enzyme';
import ListEmailTemplates from '../ListEmailTemplates';
import listOfTemplates from '../../../views/ReminderSetup/__mocks__/ReminderSetup';

describe('<ListEmailTemplates />', () => {
  const {templates, pagination} = listOfTemplates;
  const props = {
    id: null,
    history: {
      push: jest.fn()
    },
    fetchTemplates: jest.fn(),
    fetchOneTemplate: jest.fn(),
    setItemToDisable: jest.fn(),
    disableEnable: false,
    listEmailTemplatesReducer: {
      pagination: {
        pageCount: 3,
        currentPage: 3
      },
      templates: [{
        id: 1,
        name: '',
        creator : {
          fullName: ''
        },
        createdAt: '2019-01-21T07:33:26.123Z',
        updatedAt: '2019-01-21T07:33:26.123Z',
        deletedAt: null,
        disabled: true,
      }]
    },
    location: {},
    openModal: jest.fn()
  };
  const wrapper =  mount(<MemoryRouter><ListEmailTemplates {...props} /></MemoryRouter>);

  it('renders without crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('calls the onclick function', ()=>{
    wrapper.find('span.template-name').first().simulate('click');
    expect(props.openModal).toHaveBeenCalled();
  });

  it('calls onPageChange function', () => {
    wrapper.find('#next-button').simulate('click');
    expect(props.fetchTemplates).toHaveBeenCalled();
  });

  it('should display a loader while the table content is being fetched', () =>{
    const wrapper = mount(
      <ListEmailTemplates
        {...{ ...props, isLoading: true }}
      />
    );
    const reminderTemplateTableLoader = wrapper.find('ReminderTemplatePlaceholder');
    const reminderTemplateTable = wrapper.find('EmailTemplatesTable');
    expect(reminderTemplateTableLoader).toHaveLength(1);
    expect(reminderTemplateTable).toHaveLength(0);
  });
});

