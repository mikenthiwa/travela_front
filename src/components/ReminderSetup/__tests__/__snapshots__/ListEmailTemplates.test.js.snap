// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListEmailTemplates /> renders without crashing 1`] = `
Array [
  <div
    className="list-templates"
>
    <div
        className="table__container"
    >
        <table
            className="mdl-data-table mdl-js-data-table readiness-table"
        >
            <thead>
                <tr>
                    <th
                        className="mdl-data-table__cell--non-numeric table__head"
                    >
                        Template Name
                    </th>
                    <th
                        className="mdl-data-table__cell--non-numeric table__head pl-sm-100d description-left"
                    >
                        <span />
                    </th>
                    <th
                        className="mdl-data-table__cell--non-numeric table__head pl-sm-100d description-left"
                    >
                        Created By
                    </th>
                    <th
                        className="mdl-data-table__cell--non-numeric table__head pl-sm-100d description-left"
                    >
                        <span />
                    </th>
                    <th
                        className="mdl-data-table__cell--non-numeric table__head"
                    >
                        Created On
                    </th>
                    <th
                        className="mdl-data-table__cell--non-numeric table__head"
                    />
                </tr>
            </thead>
            <tbody
                className="table__body"
            >
                <tr
                    className="table__rows off"
                >
                    <td
                        className="mdl-data-table__cell--non-numeric table__data "
                    >
                        <span
                            className="template-name"
                            onClick={[Function]}
                            role="presentation"
                        />
                        <i
                            className="tiny material-icons"
                            onClick={[Function]}
                            role="presentation"
                        >
                            error
                        </i>
                    </td>
                    <td
                        className="mdl-data-table__cell--non-numeric table__data"
                    >
                        <span />
                    </td>
                    <td
                        className="mdl-data-table__cell--non-numeric table__data"
                    />
                    <td
                        className="mdl-data-table__cell--non-numeric table__data"
                    >
                        <span />
                    </td>
                    <td
                        className="mdl-data-table__cell--non-numeric table__data"
                    >
                        21-01-2019
                    </td>
                    <td
                        className="table__data"
                    >
                        <span>
                            <i
                                className="fa fa-ellipsis-v on"
                                id="toggleIcon"
                                onClick={[Function]}
                                role="presentation"
                            />
                            <div
                                className="table__menu-container "
                            >
                                <ul
                                    className="table__menu-list"
                                >
                                    <li
                                        className="table__menu-list-item bottom"
                                        id="setItem"
                                        onClick={[Function]}
                                        role="presentation"
                                    >
                                        <span
                                            className="enable"
                                        >
                                            Enable
                                        </span>
                                    </li>
                                </ul>
                            </div>
                        </span>
                    </td>
                </tr>
            </tbody>
        </table>
    </div>
</div>,
  <div>
    <div />
    <div
        className="reminder-email-templates"
    >
        <div
            className="pagination"
        >
            <button
                className="pagination__button"
                disabled={false}
                id="previous-button"
                onClick={[Function]}
                type="button"
            >
                <span
                    className=""
                >
                    Previous
                </span>
                <span
                    className=" mdl-cell--hide-desktop mdl-cell--hide-tablet"
                >
                    Previous
                </span>
            </button>
            <div
                className="pagination__items"
            >
                <span
                    className="pagination__page"
                >
                    Page
                </span>
                <div
                    className="pagination__current-page"
                    id="current-page"
                >
                    3
                </div>
                of
                <span
                    className="pagination__all-pages"
                    id="page-count"
                >
                    3
                </span>
            </div>
            <button
                className="pagination__button"
                disabled={true}
                id="next-button"
                onClick={[Function]}
                type="button"
            >
                <span
                    className=""
                >
                    Next
                </span>
                <span
                    className=" mdl-cell--hide-desktop mdl-cell--hide-tablet"
                >
                    Next
                </span>
            </button>
        </div>
    </div>
</div>,
]
`;
